src/platform/packages/shared/kbn-esql-ast/src/definitions/utils/validation/function.ts

```diff
@@ -155,11 +153,21 @@ class FunctionValidator {
     }
 
     // Validate column arguments
-    const columnMessages = this.fn.args.flat().flatMap((arg) => {
-      if (isColumn(arg) || isIdentifier(arg)) {
-        return new ColumnValidator(arg, this.context, this.parentCommand.name).validate();
+    const columnsToValidate = [];
+    const flatArgs = this.fn.args.flat();
+    for (let i = 0; i < flatArgs.length; i++) {
+      const arg = flatArgs[i];
+      if (
+        (isColumn(arg) || isIdentifier(arg)) &&
+        !(this.definition.name === '=' && i === 0) && // don't validate left-hand side of assignment
+        !(this.definition.name === 'as' && i === 1) // don't validate right-hand side of AS
```

Comment:
These cases used be covered because the user-defined columns from ALL commands were being computed in a first pass and _then_ validation was run with all those columns available in the list to check against.

Now we correctly populate the column list which means that, at the time of assignment, the new columns don't exist yet.